sdk/metric/pipeline_test.go

```diff
@@ -518,3 +536,91 @@ func TestExemplars(t *testing.T) {
 		check(t, r, 2, 2, 2)
 	})
 }
+
+func TestAddingAndObservingMeasureConcurrentSafe(t *testing.T) {
+	exp := &fnExporter{}
+	r1 := NewPeriodicReader(exp, WithInterval(10*time.Millisecond))
+	r2 := NewPeriodicReader(exp, WithInterval(10*time.Millisecond))
+
+	mp := NewMeterProvider(WithReader(r1), WithReader(r2))
+	m := mp.Meter("test")
+
+	oc1, err := m.Int64ObservableCounter("int64-observable-counter")
+	require.NoError(t, err)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := m.Int64ObservableCounter("int64-observable-counter-2")
+		require.NoError(t, err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := m.RegisterCallback(
+			func(_ context.Context, o metric.Observer) error {
+				o.ObserveInt64(oc1, 2)
+				return nil
+			}, oc1)
+		require.NoError(t, err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_ = mp.pipes[0].produce(context.Background(), &metricdata.ResourceMetrics{})
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_ = mp.pipes[1].produce(context.Background(), &metricdata.ResourceMetrics{})
+	}()
+
+	wg.Wait()
+}
+
+func TestPipelineWithMultipleReaders(t *testing.T) {
+	exp := &fnExporter{}
+	r1 := NewPeriodicReader(exp, WithInterval(10*time.Millisecond))
+	r2 := NewPeriodicReader(exp, WithInterval(10*time.Millisecond))
+
+	mp := NewMeterProvider(WithReader(r1), WithReader(r2))
+	m := mp.Meter("test")
+
+	var val atomic.Int64
+	val.Add(1)
+	measure := func(_ context.Context, m metric.Meter) {
+		oc, err := m.Int64ObservableCounter("int64-observable-counter")
+		require.NoError(t, err)
+		_, err = m.RegisterCallback(
+			// SDK periodically calls this function to collect data.
+			func(_ context.Context, o metric.Observer) error {
+				o.ObserveInt64(oc, val.Load())
+				return nil
+			}, oc)
+		require.NoError(t, err)
+	}
+	ctx := context.Background()
+	measure(ctx, m)
+	rm := new(metricdata.ResourceMetrics)
+	val.Add(1)
+	err := r1.Collect(ctx, rm)
+	require.NoError(t, err)
+	assert.EventuallyWithT(t, func(c *assert.CollectT) {
+		assert.Len(t, rm.ScopeMetrics, 1)
+		assert.Len(t, rm.ScopeMetrics[0].Metrics, 1)
```

Comment:
This will panic if previous assertion fails. I think we can simply use `require` instead of `assert`.

Reply:
See https://github.com/open-telemetry/opentelemetry-go/pull/5900/files#r1832185449